tests/datasets/storages/processors/test_replaced_groups.py

```diff
@@ -32,17 +33,68 @@ def build_in(project_column: str, projects: Sequence[int]) -> Expression:
     )
 
 
+def build_time_range(query_from: datetime, query_to: datetime) -> Expression:
+    return build_and(
+        FunctionCall(
+            None,
+            "greaterOrEquals",
+            (Column(None, None, "timestamp"), Literal(None, query_from)),
+        ),
+        FunctionCall(
+            None, "less", (Column(None, None, "timestamp"), Literal(None, query_to)),
+        ),
+    )
+
+
+def build_and(expression_one: Expression, expression_two: Expression) -> Expression:
+    return FunctionCall(None, BooleanFunctions.AND, (expression_one, expression_two))
+
+
 @pytest.fixture
 def query() -> ClickhouseQuery:
     return ClickhouseQuery(
         Table("my_table", ColumnSet([])), condition=build_in("project_id", [2]),
     )
 
 
+@pytest.fixture
+def query_with_timestamp() -> ClickhouseQuery:
+    return ClickhouseQuery(
+        Table("my_table", ColumnSet([])),
+        condition=build_and(
+            build_in("project_id", [2]),
+            build_time_range(datetime(2021, 1, 1), datetime(2021, 1, 2)),
+        ),
+    )
+
+
+@pytest.fixture
+def query_with_future_timestamp() -> ClickhouseQuery:
+    return ClickhouseQuery(
+        Table("my_table", ColumnSet([])),
+        condition=build_and(
+            build_in("project_id", [2]),
+            build_time_range(
+                datetime.now() + timedelta(days=1), datetime.now() + timedelta(days=2)
+            ),
+        ),
+    )
+
+
 def teardown_function() -> None:
     redis_client.flushdb()
 
 
+def test_query_set_final(query: ClickhouseQuery) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", None)
+
+    enforcer._set_query_final(query, True)
+    assert query.get_from_clause().final
+
+    enforcer._set_query_final(query, False)
+    assert not query.get_from_clause().final
```

Comment:
No need to test what is not a public method. 
Private/protected methods are considered internal implementation details. 
You generally rely on unit test to ensure no regression if you do a refactoring that keeps the interface unchanged. 
You run the test as it is and it must pass after the refactoring as well as before.

If your unit test tests individual private/protected methods, that cannot be done, because you will have to make changes to the test if you refactor the internals of the class. 
And when you need to make changes to the test to test for regressions you may not have introduced regressions in the test itself.